lib/db/observed.go

```diff
@@ -0,0 +1,177 @@
+// Copyright (C) 2020 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package db
+
+import (
+	"time"
+
+	"github.com/syncthing/syncthing/lib/protocol"
+)
+
+func (db *Lowlevel) AddOrUpdatePendingDevice(device protocol.DeviceID, name, address string) {
+	//FIXME locking? m.mut.Lock()
+	//FIXME locking? defer m.mut.Unlock()
+
+	key := db.keyer.GeneratePendingDeviceKey(nil, device[:])
+	od := ObservedDevice{
+		Time:    time.Now().Round(time.Second),
+		Name:    name,
+		Address: address,
+	}
+	bs, err := od.Marshal()
+	if err == nil {
+		err = db.Put(key, bs)
+	}
+	if err != nil {
+		l.Warnf("Failed to store pending device entry: %v", err)
+	}
+}
+
+// List unknown devices that tried to connect.  As a side-effect, any
+// invalid entries are dropped from the database after a warning log
+// message.  That's the only possible "repair" measure and appropriate
+// for the importance of pending entries.  They will come back soon if
+// still relevant.
+func (db *Lowlevel) PendingDevices() (map[protocol.DeviceID]ObservedDevice, error) {
+	iter, err := db.NewPrefixIterator([]byte{KeyTypePendingDevice})
+	if err != nil {
+		return nil, err
+	}
+	res := make(map[protocol.DeviceID]ObservedDevice)
+	defer iter.Release()
+	for iter.Next() {
+		bs, err := db.Get(iter.Key())
+		if err != nil {
+			return nil, err
+		}
+		var od ObservedDevice
+		deviceID := db.keyer.DeviceFromPendingDeviceKey(iter.Key())
+		//FIXME: DeviceIDFromBytes() panics when given a wrong length input.
+		//       It should rather return an error which we'd check for here.
+		if len(deviceID) != protocol.DeviceIDLength {
+			goto deleteKey
+		}
+		if err := od.Unmarshal(bs); err != nil {
+			goto deleteKey
+		}
+		res[protocol.DeviceIDFromBytes(deviceID)] = od
+		continue
+	deleteKey:
+		l.Warnln("Invalid pending device entry, deleting from database")
+		if err := db.Delete(iter.Key()); err != nil {
+			return nil, err
+		}
+	}
+	return res, nil
+}
+
+func (db *Lowlevel) RemovePendingDevice(device protocol.DeviceID) {
+	key := db.keyer.GeneratePendingDeviceKey(nil, device[:])
+	if err := db.Delete(key); err != nil {
+		l.Warnf("Failed to remove pending device entry: %v", err)
```

Comment:
I read up a bit on error handling in Go and thought that just passing up error objects without further ado is considered bad practice.  Though I did try to stick with the existing code style.

In this specific case, there is only one user (the model, especially `CommitConfiguration()`) and handling the error there would clutter the actual logic. So I opted to just log it and continue, what else could you do if `Delete()` fails?

I'm not arguing against returning `error`, but can't fully estimate the consequences yet.

Reply:
Yeah right, I guess this comes down to the same argument we had before: The importance of this data is low enough to go with just logging a warning and going on with our lives.

About your general point about error handling: Unless you are talking about adding context to errors, I do not know about what is considered bad practice about passing up errors. Some pointers about/to what you read would be nice.

Reply:
Frankly, I have no idea how to properly *handle* the error in the calling context.  With handling, I mean taking countermeasures to remedy what has failed.  That's pretty much impossible for a failed "forget this piece of information" instruction.  Especially aborting `CommitConfiguration()` and passing the error up *its* caller chain would definitely be overkill.

Probably I mixed it up, [the](https://blog.golang.org/error-handling-and-go) [articles](https://blog.golang.org/go1.13-errors) were talking about passing up errors without adding context.  Still I think just handing responsibility to the caller (even if that just means logging for future diagnostics) should only be done when there is a) no possibility to recover right away *and* b) nothing the caller could sensibly do either.

The concept of exceptions is rather nice in that regard, giving the chance to catch them and act where it is appropriate, with an included stack trace to pinpoint the full calling chain (context).  However the way Go does it forces the developer to think about the "unexpected" cases more often, so that's also nice.  Maybe its just the learning curve...

Reply:
TL;DR; I could add an error return value to `RemovePendingDevice()`, but in `RemovePendingFolder()` its already unwise to break from the loop on the first error, and otherwise it's unclear which iteration's error should be reported.  First? Last? Only a boolean for success?

Please indicate if you're convinced or what functions must have error return values ;-)

Reply:
With 

> Yeah right, I guess this comes down to the same argument we had before: The importance of this data is low enough to go with just logging a warning and going on with our lives.

I meant that I am ok with handling it as you do now (just log a warning).